src/metabase/logger.clj

```diff
@@ -52,3 +55,63 @@
       (doseq [[_ ^LoggerConfig logger-config] (.getLoggers config)]
         (.addAppender logger-config appender level filter))
       (.updateLoggers ctx))))
+
+;;; [[clojure.tools.logging]] should be using our custom Logging factory which performs NICELY and memoizes calls to
+;;; `.getLogger` instead of fetching it every single time which is SLOW in Multi-Release JARs. See #16830
+
+(defn- log-level-keyword->Level
+  "Return the Log4j2 [[Level]] that corresponds to a given [[clojure.tools.logging]] `level-keyword`."
+  ^Level [level-keyword]
+  (case level-keyword
+    :off   Level/OFF
+    :trace Level/TRACE
+    :debug Level/DEBUG
+    :info  Level/INFO
+    :warn  Level/WARN
+    :error Level/ERROR
+    :fatal Level/FATAL))
+
+;;; Extend [[org.apache.logging.log4j.Logger]] so it works with [[clojure.tools.logging]].
+;;;
+;;; Yes, https://github.com/clojure/tools.logging/blob/master/src/main/clojure/clojure/tools/logging/impl.clj#L178-L208
+;;; is presumably doing this already but it might not be a good idea to assume that code gets called. It seems like
+;;; [[clojure.tools.logging] tries to use the SLF4J logging factory by default rather than the Log4j2 factory. If that
+;;; happens then I think those lines don't get called at all. So better safe than sorry I guess.
+(extend-protocol log.impl/Logger
+  Logger
+  (enabled? [^Logger this level-keyword]
+    (.isEnabled this (log-level-keyword->Level level-keyword)))
+
+  (write! [^Logger this level-keyword ^Throwable e ^Object message]
+    (let [level (log-level-keyword->Level level-keyword)]
+      (if e
+        (.log this level message e)
+        (.log this level message)))))
+
+(def ^:private ^{:arglists '(^org.apache.logging.log4j.Logger [a-namespace])} ns-logger
+  "Get the appropriate [[Logger]] to use for a Clojure namespace. `a-namespace` can be either
+  a [[clojure.lang.Namespace]], a [[clojure.lang.Symbol]], or a [[String]]. This is memoized because `.getLogger` can
+  be pretty slow and it should never change in a prod run anyway.
+
+  When changing loggers at runtime e.g. in [[metabase.test.util.log]] you need to call [[clear-memoized-ns-loggers!]]
+  so the logger factory will pick them up."
+  (memoize/memo
+   (fn [a-namespace]
+     (.getLogger (LogManager/getContext false) (str a-namespace)))))
```

Comment:
Wouldn't it be fast enough memoizing the call with the normalized (i.e., string) `a-namespace`?